### ISSUE TITLE
Mock responses / status / user data

### ISSUE THREAD

**Issue by dimitriy-k** [damienbod/angular-auth-oidc-client#505]
How can I mock / fake login status of the user? I am using Cypress for ui tests, and of cause I would like to mock everything. 
Tried to mock some request, but no success. 
In the bad case scenario tried just to login on Identify Server with test user, but Identity Server shows empty page when you visit it from cypress (anti bots security).
Tried to return "true" in guard when for example some cookie is set. But in this case `.getIsAuthorized()` and `.getUserData()` returns still nothing.

Any help is appreciated.

**Comment by damienbod** [comment]
Cypress has a problem, in which it cannot support redirects for UI tests. We used Selenium which supports the security redirects and can test with secure SPA UI applications.

**Comment by damienbod** [comment]
See the new tests in Version 11.

**Comment by plambhsa** [comment]
@dimitriy-k did you ever figure out a way to be able to use Cypress and this auth package?

**Comment by plambhsa** [comment]
@damienbod where can I see the tests for Version 11? Do they work with Cypress now?

**Comment by dimitriy-k** [comment]
@plambhsa I use local storage variable to skip all auth stuff, I use it for local development with mocks as well for cypress tests.

**Comment by valdian** [comment]
@plambhsa @dimitriy-k This is how I login a user to IdentityServer in our cypress tests and it works for me:
```js
login(username: string, password: string) {
  const identityServerUrl = 'https://my-identityserver4-url.com';
  cy.request({
    url: identityServerUrl,
    followRedirect: false,
  }).then((response) => {
    const html = document.createElement('html');
    html.innerHTML = response.body;
    const loginForm = html.getElementsByTagName('form')[0];
    const requestVerificationToken =
      loginForm.elements['__RequestVerificationToken'].value;

    cy.request({
      method: 'POST',
      url: identityServerUrl,
      followRedirect: false,
      form: true,
      body: {
        UsernameOrEmail: username,
        Password: password,
        __RequestVerificationToken: requestVerificationToken,
      },
    });
  });
}
```

**Comment by pjlamb12** [comment]
@valdian I tried this but was running into issues where the response.body after the initial cy.request promise came back is null. So then the login form can't be found. Any ideas?

**Comment by valdian** [comment]
@pjlamb12 Check response.code or other [properties](https://docs.cypress.io/api/commands/request.html#Yields) of response, I assume you get an error.

**Comment by pjlamb12** [comment]
@valdian yeah when I checked the response with `followRedirects` set to false, all that comes back is an empty response with a 302 status code. Following the redirect seems to at least come back to the app, and doesn't die in the `beforeEach` but the authentication is not good.

**Comment by valdian** [comment]
@pjlamb12 Ok, maybe I should have commented my code snippet a little.

When I write a cypress test for content for authenticated users, I do this:

1. I call the `login` function from the code snippet I posted. `identityServerUrl` is the base URL of the Identity server (no subpath, no query parameters). The result of this method is, that I am logged in to Identity Server in the browser (i.e. an authentication cookie is set). No redirects to the application under test so far.
2. I visit a part of my app which I want to test and everything else happens automatically. The route is protected by an authentication guard, which calls this.oidcSecurityService.authorize(), when the app does not have valid tokens. There is a redirect to /connect/authorize of the Identity Server and then a redirect back to my app with authorization code in the query string because the user is already logged in to the Identity Server.

Do you set the base URL of your Identity server to `identityServerUrl`? Since you get 302 from your initial request, I assume you requested something like /connect/authorize or /connect/token.

**Comment by hylleonline** [comment]
Hi there

I know this issue is closed, but maybe someone can help anyway.

I have a very similar setup as @valdian describes. I use the same login-command in cypress. The only difference is that my identityServerUrl is <my-identityserver4-url>/Account/Login => this is where the login form resides.

My problem looks like this:
![image](https://user-images.githubusercontent.com/3962152/101021276-00be1a80-3570-11eb-9523-b862719d238f.png)

The "http://localhost:1088/ng/resolve" reports that there is no authentication. 

When I login manually I am redirected to a URL looking like this: http://localhost:1088/ng/resolve#id_token=<id_token>&access_token=<access_token>&token_type=Bearer&expires_in=1209600&scope=openid%20profile&state=<state>&session_state=<session_state>

But when running from Cypress no redirection to that URL is happening and no localstorage-values are set and thus it doesn't get authenticated.

**Comment by Arashkalantar** [comment]
در تاریخ جمعه ۱۱ سپتامبر ۲۰۲۰،‏ ۱:۳۸ ق.ظ valdian <notifications@github.com>
نوشت:

**Comment by valdian** [comment]
Hi @haidelber and sorry for the late response.
After a successful login to Identity Server, I do not visit the callback URL (/ng/resolve in your case), but I visit a route protected by my auth guard. This guard allows the user to visit the route if he is authenticated, and calls `this.oidcSecurityService.authorize()`, if there is no authenticated user. After calling the login command, the user is logged-in in Identity Server, but not in the Angular app, so the guard triggers the authentication process.

If it is not clear, how I mean it, just let me know, and I can create a sample repo to demonstrate the idea.

**Comment by wolfpackt99** [comment]
Cypress is working with specific Auth Providers that likely have client-side libraries for auth. They are working together to provide automation solutions. Would love to see support for cypress and other programmatic authentication added to this project. Cypress and this library works fine finding the input boxes and clicking submit when the client setup for test that doesn't require two-factor auth.

[Programmatic Authentication with Cypress](https://youtu.be/V-o8WzlwKmM?t=2033)

Here in the cypress realworld-app you can see the custom login commands they've written against 3 providers.
[Real World App - Auth Examples](https://github.com/cypress-io/cypress-realworld-app/tree/develop/cypress/support/auth-provider-commands)
Your personal app has to be configured to know we are in testing mode.
I think we would have to understand more about the cookies and session values that angular-auth-oidc-client was trying to set.

**Comment by rafaarcangel** [comment]
I know that this issue is close, but if this could help anyone, that's my current simulated login over an ADD.
I tried to simulate the login with the ADD without the need to open the external window to authenticate the user on microsoft.

So, at the end, I create my custom login command on Cypress, to log-in. After that, I simulate the login response, overriding the values for the storage. Here is the code:
```
Cypress.Commands.add('login', () => {
    //mechanicalrock.github.io/2020/05/05/azure-ad-authentication-cypress.html
    https: cy.request({
        method: 'POST',
        url: `https://login.microsoftonline.com/${
            Cypress.config('authentication').tenantId
        }/oauth2/v2.0/token`,
        form: true,
        body: {
            grant_type: 'password',
            client_id: Cypress.config('authentication').clientId,
            client_secret: Cypress.config('authentication').clientSecret,
            scope: Cypress.config('authentication').scopes,
            username: Cypress.config('authentication').email,
            password: Cypress.config('authentication').password,
        },
    }).then((response) => {
        const authOptions = Cypress.config('authentication');
        let token = response.body;
        var session = {
            authnResult: token,
            authzData: token.access_token,
            userData: Constants.DEFAULT_TEST_DATA,
        };

        window.sessionStorage.setItem(`0-${authOptions.clientId}`, JSON.stringify(session));
        cy.dispatch(getUserSuccess(Constants.DEFAULT_TEST_USER));
    });
});
```
Then, for my test, I created another command that send cypress to an authenticated page:
```
Cypress.Commands.add('navigateToPage', (page) => {
    sessionStorage.clear();
    cy.login();

    // -- whatever you need to add here...

    cy.visit(`${Cypress.env('CYPRESS_BASE_URL')}/${page}`);
});
```
With this approach, you can store the result in the sessionStorage, and enable oidc-client to get the data from it.

I hope this can help you!
